### PR TITLE
test: differential test between our lib and balancer's

### LINKED ISSUE #70: test(LogExpMath): Differential Fuzz LogExpMath with Balancer source
To ensure that our changes (`unchecked`) did not create divergences from original behavior.

Scope:
- LogExpMath.sol
- (potentially) LogCompression.sol
- (potentially) StableOracleMath.sol

### PR DESCRIPTION
**Pull request by xenide**
## Solution

- while writing the tests, I realized that we could do an optimization by doing `unchecked` arithmetic on `LogCompression`. It saves quite some gas, as seen in the snapshots. And since we have shown in the fuzz tests that the behaviors are equivalent, this change should be good

### CODE REVIEW

**Comment by OliverNChalk** [comment]
Yeah if Balancer is using a different solidity version, your best bet is to create the wrapper in their repo so you can getCode it and deploy it in our land

File: src/libraries/LogCompression.sol [hunk 1]
```diff
@@ -39,18 +39,22 @@ library LogCompression {
...
      */
     function toLowResLog(uint256 value) internal pure returns (int256) {
+    unchecked {
         int256 ln = LogExpMath.ln(int256(value));
 
         // Rounding division for signed numerator
         int256 lnWithError = (ln > 0 ? ln + _HALF_LOG_COMPRESSION_FACTOR : ln - _HALF_LOG_COMPRESSION_FACTOR);
         return lnWithError / _LOG_COMPRESSION_FACTOR;
     }
+    }
```
Comment (OliverNChalk) [inline src/libraries/LogCompression.sol:50]: gas optimization or causing test failures?

File: test/differential/LogCompression.t.sol [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        // we use getDeployedCode since the library contract is stateless
+        bytes memory lBytecode = vm.getDeployedCode("./reference/balancer-v2-monorepo/pkg/solidity-utils/artifacts/contracts/helpers/LogCompressionWrapper.sol/LogCompressionWrapper.json");
+        vm.etch(_balancerLogCompression, lBytecode);
+    }
+
+    function testToLowResLog(uint256 aValue) external
+    {
+        // act
+        (bool lBool, bytes memory lRes) = _balancerLogCompression.staticcall(abi.encodeWithSignature("toLowResLog(uint256)", aValue));
```
Comment (OliverNChalk) [inline test/differential/LogCompression.t.sol]: `lBool` is an example of hungarian variable names. Linus says these are only to be used by the brain damaged. As we are both smart cookies: `lBool` -> `lSuccess`

**Review by OliverNChalk — APPROVED** [review]
Looks really sick, I love it.

File: src/libraries/LogCompression.sol (hunk 1 above)
Comment (xenide) [reply src/libraries/LogCompression.sol:50]: optimization, as written above in my PR description... no failure....

File: test/differential/LogCompression.t.sol (hunk 2 above)
Comment (xenide) [reply test/differential/LogCompression.t.sol]: lol okay yeah forgot about that again. Fixed

**Pull request merged**